design/addon-categories.md

```diff
@@ -0,0 +1,140 @@
+- "PVC Provisioner"
+- "Object Store"
+- "Ingress"
+- "Logs"
+- "Snapshots"
+- "Metrics & Monitoring"
+- "Cluster Administration"
+- "Application Management"
+
+Each add-on can specify dependencies on other categories.
+
+Add-ons belonging to the same category conflict with each other.
+
+## Detailed Design
+
+Add a GET /add-ons endpoint that returns a list of add-ons along with their dependencies, what categories the add-on belongs to, and any other add-ons that are recommended to enable all features.
+Example response:
+
+```
+[
+	{
+		name: "docker",
+		fulfills: ["Container Runtime"],
+		requires: [],
+		recommends: [],
+	},{
+		name: "containerd",
+		fulfilles: ["Container Runtime"],
+		requires: [],
+		recommends: [],
+	},{
+		name: "weave",
+		fulfills: ["CNI Plugin"],
+		requires: ["Container Runtime"],
+		recommends: [],
+	},{
+		name: "calico",
+		fulfills: ["CNI Plugin"],
+		requires: ["Container Runtime"],
+		recommends: [],
+	},{
+		name: "rook",
+		fulfills: ["PVC Provisioner", "Object Store"],
+		requires: ["CNI Plugin", "Container Runtime"],
+		recommends: ["ekco"],
+	},{
+		name: "openEBS",
+		fulfills: ["PVC Provisioner"],
+		requires: ["CNI Plugin", "Container Runtime"],
+		recommends: [],
+	},{
+		name: "minio",
+		fulfills: ["Object Store"],
+		requires: ["CNI Plugin", "Container Runtime"],
+		recommends: [],
+	},{
+		name: "contour",
+		fulfills: ["Ingress"],
+		requires: ["CNI Plugin", "Container Runtime"],
+		recommends: [],
+	},{
+		name: "fluentd",
+		fulfills: ["Logs"],
+		requires: ["CNI Plugin", "Container Runtime"],
+		recommends: [],
+	},{
+		name: "kotsadm",
+		fulfills: ["Application Management"],
+		requires: ["CNI Plugin", "Container Runtime", "PVC Provisioner", "Object Store"],
+		recommends: ["registry", "velero"],
+	},{
+		name: "registry",
+		fulfills: ["Registry"],
+		requires: ["CNI Plugin", "Container Runtime","Object Store"],
+		recommends: [],
+	},{
+		name: "velero",
+		fulfills: ["Snapshots"],
+		requires: ["CNI Plugin", "Container Runtime", "Object Store"],
+		recommends: [],
+	},{
+		name: "prometheus",
+		fulfills: ["Metrics & Monitoring"],
+		requires: ["CNI Plugin", "Container Runtime", "PVC Provisioner"],
+		recommends: [],
+	},{
+		name: "ekco",
+		fulfills: ["Cluster Administration"],
+		requires: ["CNI Plugin", "Container Runtime"],
+		recommends: [],
+	}
+]
+```
+
+Each addon root directory may have a categories.json file that defines an object representing the add-on.
+The addon/<version> directories do not have separate categories.json files.
+Add a Node script to bin/ that will read all the categories.json files, concatenate them into an array, and output the result as json.
+In the deploy-staging GitHub workflow, call the Node script, save the output to a tmp file, and upload the output to the S3 staging folder.
+Do the same for the prod script.
+Clients will then be able to GET the file from either https://kurl.sh/dist/add-ons.json or https://kurl-sh.s3.amazonaws.com/dist/add-ons.json.
```

Comment:
@areed How can we ensure that these files don't get cached causing kurl.sh to be missing newly added add-ons?

Reply:
I should also note that this script actually generates two files. One for the addons and one which is the supported versions for each of the addons. The supported versions is used on both the addons page and on the create installer page.

Reply:
S3 automatically adds ETag and Last-Modified headers when the files are uploaded.